trio/_subprocess.py

```diff
@@ -214,10 +263,14 @@ def returncode(self):
         if self.poll() is None:
             async with self._wait_lock:
                 if self.poll() is None:
-                    await wait_child_exiting(self)
-                    self._proc.wait()
+                    if self._pidfd is not None:
+                        await trio.hazmat.wait_readable(self._pidfd)
+                    else:
+                        await wait_child_exiting(self)
+                    self.poll()
```

Comment:
This is changing from `self._proc.wait()` to `self.poll()`. Any particular reason? I seem to recall that using `wait()` here was necessary for correctness on OS X, because kqueue and wait() aren't quite synchronized with each other, such that the waitpid(WNOHANG) used by poll() might return zero even though kqueue said "exiting" already.

Reply:
Oof, good catch, I'd forgotten about that `waidpid(WNOHANG)` synchronization thing.

The idea was to centralize the logic for cleaning up the pidfd inside our `poll` method. But you're right, this needs a different approach.